service/src/event/service.rs

```diff
@@ -307,6 +402,71 @@ pub enum InvalidItem {
     },
 }
 
+#[async_trait]
+impl Store for CeramicEventService {
+    async fn local_sourced_data_events(&self) -> anyhow::Result<Vec<AnchorRequest>> {
+        // We must have a node DID to use here. Return an error if we don't.
+        let node_did = self
+            .node_did
+            .clone()
+            .ok_or_else(|| Error::new_app(anyhow!("node DID required to get anchor requests")))?;
+
+        // Fetch event CIDs from the events table using the previous high water mark
+        Ok(
+            CeramicOneEvent::unanchored_events_from_high_water_mark(
+                &self.pool, node_did, 1_000_000,
+            )
+            .await
+            .map_err(|e| Error::new_app(anyhow!("could not fetch unanchored events: {}", e)))?,
+        )
+    }
+
+    async fn put_time_events(&self, batch: TimeEventBatch) -> anyhow::Result<()> {
+        let proof_cid = batch.proof.to_cid()?;
+        info!(
+            "store anchor batch: proof={}, events={}",
+            proof_cid,
+            batch.time_events.events.len()
+        );
+        let events = batch
+            .time_events
+            .events
+            .iter()
+            .map(|(anchor_request, time_event)| {
+                self.build_time_event_insertable(
+                    &proof_cid,
+                    batch.proof.clone(),
+                    &batch.proof.root,
+                    time_event,
+                    anchor_request,
+                    &batch.merkle_nodes,
+                )
+            })
+            .filter_map(|r| match r {
+                Ok(e) => Some(e),
+                Err(e) => {
+                    error!("error processing time event: {}", e);
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        // TODO: We need to handle high water mark updates when there are errors here
+        CeramicOneEvent::insert_many(&self.pool, events.iter()).await?;
```

Comment:
I'm concerned about this transaction.. is there any way we can safely break this up or are we just going to take the lock for as long as it takes? We should try to benchmark this because it's what we normally expect per hours happening in one tx.

It seems like we could iter in batches and update the highwater mark after each batch? not sure if somehow erroring in the middle and having half the batch split causes a different problem though

Reply:
Yeah, that's been a concern of mine from the start, both for reads and writes.

We should write a batch then update the HWM, and so on, until were through the list. If any batch fails, the HWM doesn't move up. We can error in the middle and redo stuff as long as we don't update the HWM unless the batch was written successfully.

Reply:
@dav1do, is it ok if we do that in a follow-up PR? I'd really like to get this PR in before I'm OOO next week. Or... we could wind it up together during the retreat lol though that means pushing it out 2 more weeks...